### ISSUE TITLE
systemd-boot fails to load Linux on ARM

### ISSUE THREAD

**Issue by WolleTD** [systemd/systemd#27371]
### systemd version the issue has been seen with

252

### Used distribution

Debian Bookworm

### Linux kernel version used

6.1.0-7-armmp

### CPU architectures issue was seen on

arm

### Component

systemd-boot

### Expected behaviour you didn't see

The system boots into Linux

### Unexpected behaviour you saw

The system does not boot into Linux

### Steps to reproduce the problem

1. Create a virtual disk image with armhf Debian Bookworm, install `systemd-boot` and run `bootctl install`
2. Try to boot the system with QEMU and EDK2-ARM (I'm currently using the build by [Linaro](http://snapshots.linaro.org/components/kernel/leg-virt-tianocore-edk2-upstream/4861/QEMU-ARM/DEBUG_GCC5/)) (with `-M virt -cpu cortex-a15`, also tried `-a7`)
3. Observe systemd-boot starting, apparently invoking EFI stub and then failing the kernel launch (see output, next line should be first actual kernel message)

Note: the `mkosi` config isn't working for ARM. I used [this](https://designprincipia.com/virtualize-uefi-on-arm-using-qemu/) and [this](https://www.willhaley.com/blog/debian-arm-qemu/) blog post to get where I am now (my base image is from somewhere else, but that's not important).

Why is this an issue of systemd-boot?
- because it works with GRUB
- it even works when GRUB chainloads systemd-boot
- it works with systemd-boot built from `main`...

Yes, it's fixed in `main`. I'm opening this issue anyway because I don't understand _why_ it is fixed and, more importantly, it doesn't look like it was fixed intentionally.

I already did my own bisect. The first working version I found is https://github.com/systemd/systemd/commit/2f3c3b0bee5534f2338439f04b0aa517479f8b76.
Precisely, only the change in [Line 2318](https://github.com/systemd/systemd/commit/2f3c3b0bee5534f2338439f04b0aa517479f8b76#diff-6cb730cd48447051a413da23b3bcbd8274163d62e4879e1599120d7fa2e197aeR2318) appears to be required.

I changed this line back and forth, built systemd-boot several times and I can consisently reproduce the behaviour. I have absolutely no clue why and would like to find out and ensure systemd-boot keeps working on ARM in later versions.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
SetUefiImageMemoryAttributes - 0x00000000FC234000 - 0x0000000000001000 (0x0000000000004000)
SetUefiImageMemoryAttributes - 0x00000000FC235000 - 0x0000000000510000 (0x0000000000020000)
SetUefiImageMemoryAttributes - 0x00000000FC745000 - 0x0000000000002000 (0x0000000000004000)
InstallProtocolInterface: 09576E91-6D3F-11D2-8E39-00A0C969723B 80719E54
InstallProtocolInterface: 4006C0C1-FCB3-403E-996D-4A6C8724E06D FE8F0D10
EFI stub: Entering in SVC mode with MMU enabled
EFI stub: Booting Linux Kernel...
EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path
EFI stub: Using DTB from configuration table
EFI stub: Exiting boot services...
SetUefiImageMemoryAttributes - 0x00000000FFDA0000 - 0x0000000000007000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x00000000FFD99000 - 0x0000000000007000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x00000000FFD92000 - 0x0000000000007000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x00000000FFD89000 - 0x0000000000009000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x00000000FFD1E000 - 0x000000000006B000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x00000000FFD18000 - 0x0000000000006000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x00000000FFD12000 - 0x0000000000006000 (0x0000000000000008)
SetUefiImageMemoryAttributes - 0x00000000FFD0B000 - 0x0000000000007000 (0x0000000000000008)
```

**Comment by medhefgo** [comment]
What is your boot setup exactly? (`bootctl status` and/or the contents of the boot entry config involved)

The output shows the kernel image is running just fine, fetching the initrd from us and then exiting boot services. So it's either a gnu-efi bug or some (unrelated) UB/memory corruption somewhere else. Hard to tell, but since main branch is working fine I don't see a big issue here.

You should build main branch with `-Db_sanitize=undefined` to make certain there isn't any undefined behavior still lurking around (ubsan support is fairly recent, so it won't work on v252).

**Comment by medhefgo** [comment]
That one-liner depends on the printf machinery which landed shortly after v252. I don't consider that backporting material (especially when x86 is not affected at all here).

**Comment by bluca** [comment]
sure, we do not want to add additional risk, if there's no easy fix then that's fine

**Comment by WolleTD** [comment]
`bootctl status`:
```
root@debuerreotype:~# bootctl status --no-pager
System:
      Firmware: UEFI 2.70 (EDK II 1.00)
 Firmware Arch: arm
   Secure Boot: disabled (setup)
  TPM2 Support: no
  Boot into FW: supported

Current Boot Loader:
      Product: systemd-boot 253-1589-gf6e94c5
     Features: ✓ Boot counting
               ✓ Menu timeout control
               ✓ One-shot menu timeout control
               ✓ Default entry control
               ✓ One-shot entry control
               ✓ Support for XBOOTLDR partition
               ✓ Support for passing random seed to OS
               ✓ Load drop-in drivers
               ✓ Support Type #1 sort-key field
               ✓ Support @saved pseudo-entry
               ✓ Support Type #1 devicetree field
               ✓ Boot loader sets ESP information
          ESP: /dev/disk/by-partuuid/d1f4b6df-c98b-4be4-a5c1-e5bd6822a905
         File: └─/EFI/BOOT/BOOTARM.EFI

Random Seed:
 Passed to OS: no
 System Token: not set
       Exists: yes

Available Boot Loaders on ESP:
          ESP: /boot/efi (/dev/disk/by-partuuid/d1f4b6df-c98b-4be4-a5c1-e5bd6822a905)
         File: ├─/EFI/systemd/systemd-bootarm.efi (systemd-boot 252.6-1)
               └─/EFI/BOOT/BOOTARM.EFI (systemd-boot 253-1589-gf6e94c5)

No boot loaders listed in EFI Variables.

Boot Loader Entries:
        $BOOT: /boot/efi (/dev/disk/by-partuuid/d1f4b6df-c98b-4be4-a5c1-e5bd6822a905)

Default Boot Loader Entry:
         type: Boot Loader Specification Type #1 (.conf)
        title: Debian GNU/Linux 12 (bookworm)
           id: debian-6.1.0-7-armmp.conf
       source: /boot/efi/loader/entries/debian-6.1.0-7-armmp.conf
     sort-key: debian
      version: 6.1.0-7-armmp
        linux: /debian/6.1.0-7-armmp/linux
       initrd: /debian/6.1.0-7-armmp/initrd.img-6.1.0-7-armmp
      options: rw root=PARTUUID=42b37a96-676e-469f-88e3-7b69bb175da6
```

Entry:
```
title      Debian GNU/Linux 12 (bookworm)
version    6.1.0-7-armmp
sort-key   debian
options    rw root=PARTUUID=42b37a96-676e-469f-88e3-7b69bb175da6
linux      /debian/6.1.0-7-armmp/linux
initrd     /debian/6.1.0-7-armmp/initrd.img-6.1.0-7-armmp
```

Precise qemu call, in case there's anything interesting:
```
qemu-system-arm \
  -M virt \
  -cpu cortex-a15 \
  -serial mon:stdio \
  -bios QEMU_EFI.fd \
  -drive file=arm-debian.raw,id=hd,format=raw,if=none \
    -device virtio-scsi-device \
    -device scsi-hd,drive=hd \
  -smp 4 \
  -m 4G \
  -netdev user,id=net0 \
    -device virtio-net-device,netdev=net0 \
  -nographic
```

Besides that: yes, I'd also came to the conclusion there's UB somewhere. But I also wasn't aware that the main branch has UBSAN support in an EFI binary, which I haven't seen before, quite interesting.
My build from main shows now UB. I also inserted some UB to validate (mostly for myself) and it is detected in my setup.

@medhefgo Thanks for pointing at the UBSAN support, solved most of my worries.

@bluca We could probably just pick https://github.com/systemd/systemd/commit/2f3c3b0bee5534f2338439f04b0aa517479f8b76. I could test that, my ARM-bootloader-building-and-testing pipline is warm now.
But I don't really like that. I'd either find the actual UB or just live with it.

Maybe I'll try backporting UBSAN to the broken commit later. Unfortunately, it's not as easy due to all the gnu-efi related changes. I'll maybe waste one or two hours on that (mostly to get to know the code and that intersting UBSAN support a little). If that's not enough, let the bug be.

**Comment by medhefgo** [comment]
I was hoping the config would contain gigantic cmdline or initrd paths, but I see I nothing that could cause this.

You got me a little curious so I quickly backported it to v252, along with vendoring gnu-efi just to see if ubsan finds something: https://github.com/medhefgo/systemd/tree/v252-ubsan

(For me it already found a benign signed int overflow in gnu-efi GUID comparisons code that I had to fix).

**Comment by WolleTD** [comment]
Very cool, thank you! However, it sadly appears to be unsuccessful, I get no error (and still no boot).

UBSAN in general is working. Although I had to actually use the overflowing integer for it to trigger. In `main`, I can just add `int a = 0x7fffffff; a += 1;` and get an error. Of course I checked debug and optimization settings.

Something wild is going on. Maybe it isn't UB but something else, not entirely unused, that got optimized away? I assume you're not currently working on running gcov in EFI? :laughing:
